File: js/render/gbuffer.js
// Off-screen render targets for the deferred lighting pipeline.
//
// The scene is rendered into a G-buffer (multiple render targets + a depth
// texture) instead of straight to the screen:
//   RT0 gAlbedo  : rgb = albedo,  a = baked face shade*AO
//   RT1 gLight   : r = skylight,  g = blocklight, b = emissive, a = material id
//   RT2 gNormal  : rgb = world-space normal (0..1 encoded), a = unused
//   depth        : DEPTH_COMPONENT24 texture (reconstructs world position)
//
// A fullscreen lighting/composite pass then reads these and writes a lit colour
// target (which shares the scene depth, so forward-blended water can still
// depth-test). Everything is RGBA8 + core WebGL2 — no float-buffer extensions.
//
// Material ids (gLight.a, stored as id/255): 0 opaque, 1 foliage, 2 emissive
// surface (torch), 3 entity, 4 water, 255 sky/background.

export const MAT = { OPAQUE: 0, FOLIAGE: 1, EMISSIVE: 2, ENTITY: 3, WATER: 4, SKY: 255 };

function makeColorTex(gl, w, h, filter) {
  const t = gl.createTexture();
  gl.bindTexture(gl.TEXTURE_2D, t);
  gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA8, w, h, 0, gl.RGBA, gl.UNSIGNED_BYTE, null);
  gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, filter);
  gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, filter);
  gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
  gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
  return t;
}

export class GBuffer {
  constructor(gl) {
    this.gl = gl;
    this.w = 0; this.h = 0;
    this.scale = 1;            // render scale (quality): internal buffers = canvas * scale
    this.sceneFBO = null;
    this.litFBO = null;        // lit colour + scene depth (forward water/selection)
    this.litColorFBO = null;   // lit colour only (composite reads scene depth as a texture)
    this.heldFBO = null;       // lit colour + a scratch depth buffer (viewmodel)
    this.gAlbedo = this.gLight = this.gNormal = this.depth = this.lit = null;
    this.heldDepth = null;
    // a pool of extra half/full screen single-target FBOs for effect passes
    this.aux = [];             // [{fbo, tex, w, h}]
  }

  // Internal buffer size (canvas size * render scale, min 1).
  bufW() { return Math.max(1, Math.round(this.w * this.scale)); }
  bufH() { return Math.max(1, Math.round(this.h * this.scale)); }

  resize(w, h, scale) {
    if (w === this.w && h === this.h && scale === this.scale && this.sceneFBO) return;
    this.w = w; this.h = h; this.scale = scale;
    const gl = this.gl;
    const bw = this.bufW(), bh = this.bufH();
    this._dispose();

    // --- scene G-buffer: 3 colour targets + depth texture ---
    this.gAlbedo = makeColorTex(gl, bw, bh, gl.NEAREST);
    this.gLight = makeColorTex(gl, bw, bh, gl.NEAREST);
    this.gNormal = makeColorTex(gl, bw, bh, gl.NEAREST);
    this.depth = gl.createTexture();
    gl.bindTexture(gl.TEXTURE_2D, this.depth);
    gl.texImage2D(gl.TEXTURE_2D, 0, gl.DEPTH_COMPONENT24, bw, bh, 0, gl.DEPTH_COMPONENT, gl.UNSIGNED_INT, null);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.NEAREST);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.NEAREST);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);

    this.sceneFBO = gl.createFramebuffer();
    gl.bindFramebuffer(gl.FRAMEBUFFER, this.sceneFBO);
    gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, this.gAlbedo, 0);
    gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT1, gl.TEXTURE_2D, this.gLight, 0);
    gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT2, gl.TEXTURE_2D, this.gNormal, 0);
    gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.DEPTH_ATTACHMENT, gl.TEXTURE_2D, this.depth, 0);
    gl.drawBuffers([gl.COLOR_ATTACHMENT0, gl.COLOR_ATTACHMENT1, gl.COLOR_ATTACHMENT2]);
    this._check("scene");

    // --- lit colour target. Two FBOs share the SAME colour texture: one carries
    // the scene depth (forward water/selection/held depth-test against it), one is
    // depth-less so the composite pass can sample the depth texture without a
    // read/write feedback loop. ---
    this.lit = makeColorTex(gl, bw, bh, gl.LINEAR);
    this.litFBO = gl.createFramebuffer();
    gl.bindFramebuffer(gl.FRAMEBUFFER, this.litFBO);
    gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, this.lit, 0);
    gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.DEPTH_ATTACHMENT, gl.TEXTURE_2D, this.depth, 0);
    gl.drawBuffers([gl.COLOR_ATTACHMENT0]);
    this._check("lit");

    this.litColorFBO = gl.createFramebuffer();
    gl.bindFramebuffer(gl.FRAMEBUFFER, this.litColorFBO);
    gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, this.lit, 0);
    gl.drawBuffers([gl.COLOR_ATTACHMENT0]);
    this._check("litColor");

    // --- viewmodel target: the lit colour again, but with a SCRATCH depth buffer.
    // The held item is drawn through its own close-up projection, so its depths
    // mean nothing against the scene's — yet it still has to hide its own back
    // faces. A private depth buffer we can clear gives the viewmodel a real
    // depth test against itself while leaving the scene depth (which the god-ray
    // and underwater passes still read) untouched. ---
    this.heldDepth = gl.createRenderbuffer();
    gl.bindRenderbuffer(gl.RENDERBUFFER, this.heldDepth);
    gl.renderbufferStorage(gl.RENDERBUFFER, gl.DEPTH_COMPONENT16, bw, bh);
    gl.bindRenderbuffer(gl.RENDERBUFFER, null);
    this.heldFBO = gl.createFramebuffer();
    gl.bindFramebuffer(gl.FRAMEBUFFER, this.heldFBO);
    gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, this.lit, 0);
    gl.framebufferRenderbuffer(gl.FRAMEBUFFER, gl.DEPTH_ATTACHMENT, gl.RENDERBUFFER, this.heldDepth);
    gl.drawBuffers([gl.COLOR_ATTACHMENT0]);
    this._check("held");

    gl.bindFramebuffer(gl.FRAMEBUFFER, null);
  }

  // A scratch single-target colour FBO at (full*frac) resolution, created on
  // demand and cached by slot index. Used by effect passes (godrays, SSR, blur).
  auxTarget(slot, frac, filter) {
    const gl = this.gl;
    const w = Math.max(1, Math.round(this.bufW() * frac));
    const h = Math.max(1, Math.round(this.bufH() * frac));
    let a = this.aux[slot];
    if (a && a.w === w && a.h === h) return a;
    if (a) { gl.deleteFramebuffer(a.fbo); gl.deleteTexture(a.tex); }
    const tex = makeColorTex(gl, w, h, filter || gl.LINEAR);
    const fbo = gl.createFramebuffer();
    gl.bindFramebuffer(gl.FRAMEBUFFER, fbo);
    gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, tex, 0);
    gl.bindFramebuffer(gl.FRAMEBUFFER, null);
    a = this.aux[slot] = { fbo, tex, w, h };
    return a;
  }

  bindScene() {
    const gl = this.gl;
    gl.bindFramebuffer(gl.FRAMEBUFFER, this.sceneFBO);
    gl.viewport(0, 0, this.bufW(), this.bufH());
  }
  bindLit() {
    const gl = this.gl;
    gl.bindFramebuffer(gl.FRAMEBUFFER, this.litFBO);
    gl.viewport(0, 0, this.bufW(), this.bufH());
  }
  bindLitColor() {
    const gl = this.gl;
    gl.bindFramebuffer(gl.FRAMEBUFFER, this.litColorFBO);
    gl.viewport(0, 0, this.bufW(), this.bufH());
  }
  // The viewmodel target, with its scratch depth cleared and ready to test.
  bindHeld() {
    const gl = this.gl;
    gl.bindFramebuffer(gl.FRAMEBUFFER, this.heldFBO);
    gl.viewport(0, 0, this.bufW(), this.bufH());
    gl.depthMask(true);
    gl.clear(gl.DEPTH_BUFFER_BIT);
  }

  _check(label) {
    const gl = this.gl;
    const s = gl.checkFramebufferStatus(gl.FRAMEBUFFER);
    if (s !== gl.FRAMEBUFFER_COMPLETE) console.error(`GBuffer ${label} incomplete: 0x${s.toString(16)}`);
  }

  _dispose() {
    const gl = this.gl;
    for (const t of [this.gAlbedo, this.gLight, this.gNormal, this.depth, this.lit]) if (t) gl.deleteTexture(t);
    if (this.heldDepth) gl.deleteRenderbuffer(this.heldDepth);
    if (this.sceneFBO) gl.deleteFramebuffer(this.sceneFBO);
    if (this.litFBO) gl.deleteFramebuffer(this.litFBO);
    if (this.litColorFBO) gl.deleteFramebuffer(this.litColorFBO);
    if (this.heldFBO) gl.deleteFramebuffer(this.heldFBO);
    this.sceneFBO = this.litFBO = this.litColorFBO = this.heldFBO = null;
    this.heldDepth = null;
  }

  dispose() {
    this._dispose();
    const gl = this.gl;
    for (const a of this.aux) if (a) { gl.deleteFramebuffer(a.fbo); gl.deleteTexture(a.tex); }
    this.aux = [];
  }
}

File: src/render/gbuffer.cpp
#include "render/gbuffer.h"

#include <algorithm>
#include <cmath>

#include "core/log.h"

namespace hr::render {
namespace {

GLuint makeColorTex(int w, int h, GLenum filter) {
  GLuint t = 0;
  glGenTextures(1, &t);
  glBindTexture(GL_TEXTURE_2D, t);
  glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, w, h, 0, GL_RGBA, GL_UNSIGNED_BYTE, nullptr);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, filter);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, filter);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
  return t;
}

}  // namespace

int GBuffer::bufW() const { return std::max(1, static_cast<int>(std::lround(w_ * scale_))); }
int GBuffer::bufH() const { return std::max(1, static_cast<int>(std::lround(h_ * scale_))); }

bool GBuffer::check(const char* label) const {
  const GLenum s = glCheckFramebufferStatus(GL_FRAMEBUFFER);
  if (s != GL_FRAMEBUFFER_COMPLETE) {
    log::error("gbuffer: %s framebuffer incomplete (0x%04X)", label, s);
    return false;
  }
  return true;
}

void GBuffer::resize(int width, int height, float scale) {
  if (width == w_ && height == h_ && scale == scale_ && sceneFbo_) return;
  w_ = width;
  h_ = height;
  scale_ = scale;
  disposeCore();

  const int bw = bufW(), bh = bufH();

  // --- scene: three colour targets plus a depth texture ---
  albedo_ = makeColorTex(bw, bh, GL_NEAREST);
  light_ = makeColorTex(bw, bh, GL_NEAREST);
  normal_ = makeColorTex(bw, bh, GL_NEAREST);

  glGenTextures(1, &depth_);
  glBindTexture(GL_TEXTURE_2D, depth_);
  glTexImage2D(GL_TEXTURE_2D, 0, GL_DEPTH_COMPONENT24, bw, bh, 0, GL_DEPTH_COMPONENT,
               GL_UNSIGNED_INT, nullptr);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);

  glGenFramebuffers(1, &sceneFbo_);
  glBindFramebuffer(GL_FRAMEBUFFER, sceneFbo_);
  glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, albedo_, 0);
  glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0 + 1, GL_TEXTURE_2D, light_, 0);
  glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0 + 2, GL_TEXTURE_2D, normal_, 0);
  glFramebufferTexture2D(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_TEXTURE_2D, depth_, 0);
  {
    const GLenum bufs[3] = {GL_COLOR_ATTACHMENT0, GL_COLOR_ATTACHMENT0 + 1,
                            GL_COLOR_ATTACHMENT0 + 2};
    glDrawBuffers(3, bufs);
  }
  check("scene");

  // --- lit colour. Two FBOs share the SAME colour texture: one carries the scene
  // depth, so forward water and the selection outline can depth-test against the
  // opaque scene; the other is depth-less, so the composite can sample the depth
  // texture without a read/write feedback loop on its own attachment. ---
  lit_ = makeColorTex(bw, bh, GL_LINEAR);

  glGenFramebuffers(1, &litFbo_);
  glBindFramebuffer(GL_FRAMEBUFFER, litFbo_);
  glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, lit_, 0);
  glFramebufferTexture2D(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_TEXTURE_2D, depth_, 0);
  {
    const GLenum bufs[1] = {GL_COLOR_ATTACHMENT0};
    glDrawBuffers(1, bufs);
  }
  check("lit");

  glGenFramebuffers(1, &litColorFbo_);
  glBindFramebuffer(GL_FRAMEBUFFER, litColorFbo_);
  glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, lit_, 0);
  {
    const GLenum bufs[1] = {GL_COLOR_ATTACHMENT0};
    glDrawBuffers(1, bufs);
  }
  check("litColor");

  // --- viewmodel: the lit colour again, but with a SCRATCH depth buffer we can
  // clear. That gives the held item a real depth test against itself while leaving
  // the scene depth — which the god-ray and underwater passes still read — intact.
  // Without any depth test a model draws in emission order and its back faces paint
  // over its front ones, which is what made a held block lose its top face. ---
  glGenRenderbuffers(1, &heldDepth_);
  glBindRenderbuffer(GL_RENDERBUFFER, heldDepth_);
  glRenderbufferStorage(GL_RENDERBUFFER, GL_DEPTH_COMPONENT16, bw, bh);
  glBindRenderbuffer(GL_RENDERBUFFER, 0);

  glGenFramebuffers(1, &heldFbo_);
  glBindFramebuffer(GL_FRAMEBUFFER, heldFbo_);
  glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, lit_, 0);
  glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_RENDERBUFFER, heldDepth_);
  {
    const GLenum bufs[1] = {GL_COLOR_ATTACHMENT0};
    glDrawBuffers(1, bufs);
  }
  check("held");

  glBindFramebuffer(GL_FRAMEBUFFER, 0);
  log::debug("gbuffer: %dx%d (scale %.2f)", bw, bh, scale_);
}

GBuffer::Aux& GBuffer::auxTarget(int slot, float frac, GLenum filter) {
  if (static_cast<int>(aux_.size()) <= slot) aux_.resize(slot + 1);
  Aux& a = aux_[slot];

  const int w = std::max(1, static_cast<int>(std::lround(bufW() * frac)));
  const int h = std::max(1, static_cast<int>(std::lround(bufH() * frac)));
  if (a.fbo && a.w == w && a.h == h) return a;

  if (a.fbo) {
    glDeleteFramebuffers(1, &a.fbo);
    glDeleteTextures(1, &a.tex);
  }
  a.tex = makeColorTex(w, h, filter);
  glGenFramebuffers(1, &a.fbo);
  glBindFramebuffer(GL_FRAMEBUFFER, a.fbo);
  glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, a.tex, 0);
  glBindFramebuffer(GL_FRAMEBUFFER, 0);
  a.w = w;
  a.h = h;
  return a;
}

void GBuffer::bindScene() {
  glBindFramebuffer(GL_FRAMEBUFFER, sceneFbo_);
  glViewport(0, 0, bufW(), bufH());
}

void GBuffer::bindLit() {
  glBindFramebuffer(GL_FRAMEBUFFER, litFbo_);
  glViewport(0, 0, bufW(), bufH());
}

void GBuffer::bindLitColor() {
  glBindFramebuffer(GL_FRAMEBUFFER, litColorFbo_);
  glViewport(0, 0, bufW(), bufH());
}

void GBuffer::bindHeld() {
  glBindFramebuffer(GL_FRAMEBUFFER, heldFbo_);
  glViewport(0, 0, bufW(), bufH());
  // glClear respects the depth write mask, so it has to be enabled first.
  glDepthMask(GL_TRUE);
  glClear(GL_DEPTH_BUFFER_BIT);
}

void GBuffer::disposeCore() {
  for (GLuint* t : {&albedo_, &light_, &normal_, &depth_, &lit_}) {
    if (*t) {
      glDeleteTextures(1, t);
      *t = 0;
    }
  }
  if (heldDepth_) {
    glDeleteRenderbuffers(1, &heldDepth_);
    heldDepth_ = 0;
  }
  for (GLuint* f : {&sceneFbo_, &litFbo_, &litColorFbo_, &heldFbo_}) {
    if (*f) {
      glDeleteFramebuffers(1, f);
      *f = 0;
    }
  }
}

void GBuffer::dispose() {
  disposeCore();
  for (Aux& a : aux_) {
    if (a.fbo) {
      glDeleteFramebuffers(1, &a.fbo);
      glDeleteTextures(1, &a.tex);
    }
  }
  aux_.clear();
}

}  // namespace hr::render
